### PR TITLE
issue 14134 -  Free of large array does not work

### PR DESCRIPTION
**Pull request by rainers**
https://issues.dlang.org/show_bug.cgi?id=14134

"delete slice" now destroys and frees the full array

### CODE REVIEW

File: src/rt/lifetime.d [hunk 1]
```diff
@@ -1193,24 +1186,52 @@ extern (C) void _d_delarray_t(void[]* p, const TypeInfo_Struct ti)
...
             }
-            GC.free((*p).ptr);
+
+            // if p is in the cache, clear it there as well
+            if(bic)
+                bic.base = null;
+
+            GC.free(info.base);
         }
         *p = null;
```
Comment (schveiguy) [inline src/rt/lifetime.d]: Should we set `*p` to `null` if it's not a GC array? How does one realize that delete didn't actually do anything?

**Comment by schveiguy** [comment]
I can live with this, but I would wonder if we should add a check to ensure the array pointer points at the start of the array block. This would be consistent with how GC.free works.

File: src/rt/lifetime.d (hunk 1 above)
Comment (rainers) [inline src/rt/lifetime.d]: Not changing p if it's not freed sounds reasonable. Would that be a change to the language specification? Unfortunately the documentation is pretty short for "delete".

**Comment by rainers** [comment]
> can live with this, but I would wonder if we should add a check to ensure the array pointer points at the start of the array block.

Yeah, this was also an option mentioned in the bug report comments. It seems only half-correct to check the start pointer, but not the length, though. Both indicate the caller might not actually "own" the array.

File: src/rt/lifetime.d (hunk 1 above)
Comment (rainers) [inline src/rt/lifetime.d]: I've moved clearing the passed array into the conditional.

**Comment by rainers** [comment]
> It seems only half-correct to check the start pointer, but not the length, though.

I just notice that we cannot always check the length, because we only have type info for structs with dtors.

**Comment by schveiguy** [comment]
> Yeah, this was also an option mentioned in the bug report comments. It seems only half-correct to check the start pointer, but not the length, though. Both indicate the caller might not actually "own" the array.

But it's very normal to truncate with `arr.length -= ...`. Slicing off the front is not as expected.

The point I was making is that we already have a requirement that GC.free will only work with a block pointer. However, at this point, I'm not too worried since `delete` is going away anyway, so I'm fine with what you have.

> I just notice that we cannot always check the length, because we only have type info for structs with dtors.

`delete` should give you the static `TypeInfo` of the element, which is all you need to see if the overall length matches what's stored in the array block.

But in any case, we probably shouldn't spend too much time on this, see previous comment ;)

**Comment by schveiguy** [comment]
> `delete` should give you the static `TypeInfo` of the element,

I take this back, I see it doesn't actually! (though it could) So yeah, no way to check that unfortunately.

**Comment by MartinNowak** [comment]
> Both indicate the caller might not actually "own" the array.

Both GC.free and delete are unsafe operations exactly because they can delete memory they don't own. Requiring that delete is only called for a "complete" array doesn't make it any safer.

File: src/rt/lifetime.d [hunk 2]
```diff
@@ -1193,24 +1186,53 @@ extern (C) void _d_delarray_t(void[]* p, const TypeInfo_Struct ti)
 {
     if (p)
     {
-        assert(!(*p).length || (*p).ptr);
-        if ((*p).ptr)
+        auto bic = __getBlkInfo(p.ptr);
+        auto info = bic ? *bic : GC.query(p.ptr);
+
+        if (info.base && (info.attr & BlkAttr.APPENDABLE))
```
Comment (MartinNowak) [inline src/rt/lifetime.d]: Why the added appendable check?

File: src/rt/lifetime.d [hunk 3]
```diff
@@ -1193,24 +1186,53 @@ extern (C) void _d_delarray_t(void[]* p, const TypeInfo_Struct ti)
 {
     if (p)
```
Comment (MartinNowak) [inline src/rt/lifetime.d]: How could that ever be null?

**Comment by schveiguy** [comment]
@MartinNowak did you accidentally merge this?

File: src/rt/lifetime.d (hunk 2 above)
Comment (schveiguy) [inline src/rt/lifetime.d]: Don't you need to know if APPENDABLE is there in order to verify the array to be deleted starts at the right point?

Comment (rainers) [inline src/rt/lifetime.d]: > Why did you add the appendable check?

Should we also free anything that was not allocated as an array by the runtime? Running dtors is dangerous in this case, though we could check flags only in case a type info is passed.

File: src/rt/lifetime.d (hunk 3 above)
Comment (rainers) [inline src/rt/lifetime.d]: It's been there so far, but it seems gratitious. We might want to convert it to an assert.

File: src/rt/lifetime.d (hunk 2 above)
Comment (MartinNowak) [inline src/rt/lifetime.d]: Not sure, but I'd tend to check APPENDABLE too, because of this.

```d
class Foo
{
    static struct Bar { ~this() {} }
    Bar[4] bars;
}

unittest
{
    auto foo = new Foo;
    auto bars = foo.bars[];
    delete bars;
}
```

**Comment by MartinNowak** [comment]
> @MartinNowak did you accidentally merge this?

No, it fixes the issue for now. But I thought we might have some more ideas on the topic.

File: src/rt/lifetime.d (hunk 2 above)
Comment (schveiguy) [inline src/rt/lifetime.d]: Hm... that doesn't seem like the right correlation.

For instance:

```D
auto foo = new Foo[2];
auto bars = foo[0].bars[];
delete bars;
```

I wonder if the type info to call destuctors shouldn't be read from the GC instead of relying on the static type. I didn't review that system that adds finalizer capability to structs. Is this what it is supposed to do?

**Comment by schveiguy** [comment]
OK, I just was confused with the lack of auto-merge ;)

File: src/rt/lifetime.d (hunk 2 above)
Comment (rainers) [inline src/rt/lifetime.d]: It could be used, but it would not allow to manually emplace structs into a byte array and delete them later:
```
auto foo = new ubyte[2*Bar.sizeof];
auto bars = cast(Bar[])foo;
delete bars;
```

Comment (schveiguy) [inline src/rt/lifetime.d]: Well, in that case, you are on your own anyway :)

```D
// delete bars;
bars = null; // let GC destroy... oops
```

Comment (rainers) [inline src/rt/lifetime.d]: Well, that would not call any destructors. I think that might be expected.

Comment (schveiguy) [inline src/rt/lifetime.d]: I don't know whether it's expected or not, but I think I might expect either possibility when I call delete.

What does make the most sense to me is to have delete be a "pre-emptive GC collection," so everything that is done via GC would be done via delete, except synchronously.

In any case, we are spending time debating a to-be-deprecated feature :) Nice to have these discussions though, for when we do actually have to create a replacement.

Comment (MartinNowak) [inline src/rt/lifetime.d]: > debating a to-be-deprecated feature

It's probably TBD since 3 years, yet people keep using it.

**Comment by WalterBright** [comment]
When writing pull requests that fix a bugzilla issue, write the title as:
```
Fix Issue NNNN ....
```
so that the bugzilla issue ill be automatically marked as fixed when it gets pulled.

**Comment by rainers** [comment]
> When writing pull requests that fix a bugzilla issue, write the title as:
> Fix Issue NNNN ....

I was not sure this is a fix, it was meant as a contribution towards a fix. Is calling `delete` going to be the proposed way to manually free an array from memory?
